src/core/apparmor-setup.c

```diff
@@ -0,0 +1,99 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <errno.h>
+#if HAVE_APPARMOR
+#include <sys/apparmor.h>
+#endif
+#include <unistd.h>
+
+#include "apparmor-setup.h"
+#include "apparmor-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "log.h"
+#include "macro.h"
+#include "string-util.h"
+#include "strv.h"
+
+#if HAVE_APPARMOR
+DEFINE_TRIVIAL_CLEANUP_FUNC(aa_kernel_interface *, aa_kernel_interface_unref);
+#endif
+
+int mac_apparmor_setup(void) {
+#if HAVE_APPARMOR
+        int r;
+        _cleanup_close_ int fd = -1;
+        _cleanup_(aa_kernel_interface_unrefp) aa_kernel_interface *interface = NULL;
+        const char *current_file, *profile_file;
+        _cleanup_free_ char *current_profile = NULL;
+
+        if (!mac_apparmor_use()) {
+                log_debug("AppArmor either not supported by the kernel or disabled.");
+                return 0;
+        }
+
+        /* To enable LSM stacking a patch to the kernel is proposed to create a
+         * per-LSM subdirectory to distinguish between the LSMs. Therefore, we
+         * read the file from the LSM specific directory first and only if that
+         * fails the one from the generic directory.
+         */
+        FOREACH_STRING(current_file, "/proc/self/attr/apparmor/current", "/proc/self/attr/current") {
+                r = read_one_line_file(current_file, &current_profile);
+                if (r == -ENOENT)
+                        continue;
+                else if (r < 0)
+                        log_warning_errno(r, "Failed to read current AppArmor profile from file %s, ignoring: %m", current_file);
+                else
+                        break;
+        }
+        if (!current_profile) {
+                log_warning("Failed to get the current AppArmor profile of systemd from /proc/self/attr/apparmor/current or /proc/self/attr/current, ignoring.");
+                return 0;
+        }
+        if (!streq(current_profile, "unconfined")) {
+                log_debug("We are already confined in an AppArmor profile.");
+                return 0;
+        }
+
+        FOREACH_STRING(profile_file, "/etc/systemd/profile.apparmor", "/usr/lib/systemd/profile.apparmor") {
+                fd = open(profile_file, O_RDONLY | O_CLOEXEC);
+                if (fd < 0) {
+                        if (errno == ENOENT)
+                                continue;
+                        log_warning_errno(errno, "Failed to open binary AppArmor profile file %s, ignoring: %m", profile_file);
+                } else
+                        break;
+        }
```

Comment:
I have concerns here both of these locations are outside of apparmor regular cache management, which means they will require manual update to keep from being stale. Also neither of these take into account that policy can be compiled differently for different kernel versions. This can lead to the following cases

1. Everything working because the policy features matches up well with the kernel.
2. The kernel rejecting policy because it uses an abi, or feature set it doesn't know how to handle.
3. The kernel accepting the policy but it doesn't enforce everything it could. The kernel is backwards compatible to older abis, but when using a single cache file and multiple kernels the policy cache has to be compiled to the lowest common denominator.

It would be far better if we can standardize this to work within regular apparmor policy compiles and loads, which would mean we could have the cache being updated when policy is updated and also different cache files for the different kernel versions.